qiskit/mapper/_layout.py

```diff
@@ -48,62 +50,70 @@ def from_list(self, input_list):
             [(QuantumRegister(3, 'qr'), 0), None,
              (QuantumRegister(3, 'qr'), 2), (QuantumRegister(3, 'qr'), 3)]
         """
-        for key, value in enumerate(input_list):
-            self[key] = value
+        for physical, virtual in enumerate(input_list):
+            if Layout.is_virtual(virtual):
+                self._set_type_checked_item(virtual, physical)
+
+    @staticmethod
+    def is_virtual(value):
+        """Checks if value has the format of a virtual qubit """
+        return value is None or isinstance(value, tuple) and \
+               len(value) == 2 and \
+               isinstance(value[0], Register) and isinstance(value[1], int)
 
     def __getitem__(self, item):
-        # N.B. Layout.__len__ is O(layout size). Short circuit early when possible.
-        if isinstance(item, int) and item not in self and item < len(self):
-            return None
-        return dict.__getitem__(self, item)
+        if item in self._p2v:
+            return self._p2v[item]
+        if item in self._v2p:
+            return self._v2p[item]
+        raise KeyError('The item %s does not exist in the Layout' % (item,))
 
     def __setitem__(self, key, value):
-        Layout._checktype(key)
-        Layout._checktype(value)
-        if isinstance(key, type(value)):
-            raise LayoutError('Key (%s) and value (%s) cannot have the same type' % (key, value))
-
-        if key in self:
-            del self[key]
-        if value in self:
-            del self[value]
-        if key is not None:
-            dict.__setitem__(self, key, value)
-        if value is not None:
-            dict.__setitem__(self, value, key)
+        if isinstance(key, int) and Layout.is_virtual(value):
+            physical = key
+            virtual = value
+        elif isinstance(value, int) and Layout.is_virtual(key):
+            physical = value
+            virtual = key
+        else:
+            raise LayoutError('The map (%s -> %s) has to be a ((Register, integer) -> integer)'
+                              ' or the other way around.' % (key, value))
+
+        self._set_type_checked_item(virtual, physical)
+
+    def _set_type_checked_item(self, virtual, physical):
+        if physical in self._p2v:
+            del self._p2v[physical]
+        if virtual in self._p2v:
+            del self._v2p[virtual]
+
```

Comment:
Is there a risk of leaving a dangling pointer here? e.g. if `layout = Layout({ (q,0): 0 })` and I `layout[(q,0)] = 1`. What will `layout[0]` return?

Reply:
Indeed. I added a test to check that out. But I have the feeling that my solution can be more efficient